rfcs/0000-cluster-signing.md

```diff
@@ -0,0 +1,109 @@
+# Allow cluster-wide image signing with `cosign`
+
+## Problem
+
+The signing solutions in `kpack` are currently all namespace or `Image`/`Build` specific. This means that it is harder
+for kpack operators to set image-signing up for all of the users of a `kpack` cluster. Given that these signing solutions
+are `namespace` specific, each user needs to set up their signing toolchain on their own which can be quite a bit of an effort.
+
+In the past we have discussed having the ability to provide cluster-wide image signing that would allow all images generated by a `kpack` cluster to be signed. This supplements the [`cosign RFC`](https://github.com/pivotal/kpack/pull/694).
+
+But the main blocker to such an implementation has been the architectural issue of signing images without cross-mounting secrets from different namespaces or leaking the cluster-wide signing key to the users. See discussions at this [Github comment thread](https://github.com/pivotal/kpack/pull/694#discussion_r631321627)
+
+`cosign` however provides an architectural design that would allow us to enable cluster wide image signing without leaking or cross mounting secrets.
+
+## Outcome
+
+`kpack` integrates with `cosign` to sign each and every image on a cluster-wide basis (if configured) and it builds and pushs the
+signatures to a registry so that users can ensure the chain of custody of a generated artifact.
+
+This proposal aims to cover specifically the flow of signing an image produced by `kpack Image` resource without verification of base or builder images pulled in the process.
+
+## Actions to take
+
+### Enable cluster-wide image signing using `cosign`
+
+- Implement a flow that generates the `cosign` signature payload for the image on the controller side based on the image digest output from the `Build`, then calculates its signature and pushes it either to the registry specified using the [to the registry specified in the `COSIGN_REPOSITORY` environment variable](#key-generation-and-storage), using the image signing credentials provided to the controller. This flow must happen after the image has been pushed to the registry.
+
+- If `cosign` fails to sign an image, the build should fail and output an error message in the build log, so the operator can troubleshoot the issue. The error messages should also be added in any other places where error messages are presented.
```

Comment:
If the signing is happening in the controller would it be acceptable to simply report the signing error on the build status?